Make terminal table more responsive

hjanuschka opened the pull request:
this trys to fix the ugliness of terminal table on small terminals - by making tables fit to terminal width.
in addition to the variable size, i also did a middle_truncate.


so instead of `very long va...`  now `very ... value`  is displayed.


# Large
![large](https://cloud.githubusercontent.com/assets/2891702/21029293/56fd96bc-bd99-11e6-9352-1028db11d62b.png)

# Medium
![medium](https://cloud.githubusercontent.com/assets/2891702/21029292/56fa35f8-bd99-11e6-9383-6f2797a934a9.png)

# Small
![small](https://cloud.githubusercontent.com/assets/2891702/21029294/56feddba-bd99-11e6-862e-519e62f081c0.png)


> Ruby 1.9.3 added 'io/console' to the standard library.
the additional require should not be a problem.

to quickly test it use this ruby script:

```ruby
require "fastlane_core"
FastlaneCore::PrintTable.print_values(config:
  { wwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwww: "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb bbbbbbbbbbbbbbbbbbbbbbb bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
      wwwwwwwwwwwwwwwwwwwwwwwwxw: "ccc bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb bbbbbbbbbbbbbbbbbbbbbbb bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb" },

                                   hide_keys: [],
                                       title: "Summary SUM")

```

hjanuschka committed:
make terminal table resposnive

KrauseFx commented on fastlane_core/lib/fastlane_core/string_filters.rb:
```diff
@@ -31,4 +31,14 @@ def truncate(truncate_at, options = {})
   end
+
+  # Base taken from: http://stackoverflow.com/a/12202205/1945875
+  def middle_truncate(length = 20, options = {})
```
Is there a good way to detect the initial terminal width?

hjanuschka replied:
@krausefx
this is just a generic function.
in the table printer itself - the terminal width is used

tcols = IO.console.winsize[1]
max_length = tcols

KrauseFx replied:
Ah yeah, cool 👍

milch commented:
This is a really good idea, however, I would also like an option to disable this (e.g. when running on CI) so that users have at least an option to see the whole thing for debugging purposes.

hjanuschka commented:
@milch  ok done - 

```ruby
truncate = false if Helper.is_ci? || FastlaneCore::Env.truthy?("DO_NOT_TRUNCATE_TABLES")
```

sorry for the force push, as i had to rebase for the `truthy?` stuff.

milch commented:
> sorry for the force push

No worries, I haven't reviewed yet 👍 

> `FastlaneCore::Env.truthy?("DO_NOT_TRUNCATE_TABLES")`

This is a bit nitpicky but mostly we'd like to prefix ENV-vars. Can we call it `FL_NO_TRUNCATE_TABLES` (or something)?

KrauseFx commented on fastlane_core/lib/fastlane_core/print_table.rb:
```diff
@@ -18,7 +18,10 @@ def print_values(config: nil, title: nil, hide_keys: [], mask_keys: [])
 
         params = {}
-        params[:rows] = limit_row_size(rows)
+        truncate = false if Helper.is_ci? || FastlaneCore::Env.truthy?("DO_NOT_TRUNCATE_TABLES")
```
Instead of `DO_NOT`, we've used `SKIP_...` until now

KrauseFx commented:
This looks super promising, ideally it would also be used for the action table: https://github.com/fastlane/fastlane/issues/7811

hjanuschka commented:
@KrauseFx feedback addressed.
should i do the action table thing in this PR or in a seperate?

hjanuschka commented:
@KrauseFx ok i updated it to also work with newlines instead of the `..` stuff (defaulting to: `....`)


here is a sample to test:

```ruby

require "fastlane_core"



FastlaneCore::PrintTable.print_values(config:
  { wwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwww: "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb bbbbbbbbbbbbbbbbbbbbbbb bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
      wwwwwwwwwwwwwwwwwwwwwwwwxw: "ccc bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb bbbbbbbbbbbbbbbbbbbbbbb bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb" },

                                   hide_keys: [],
                                       title: "Summary SUM")


FastlaneCore::PrintTable.print_values(config:
                                         { wwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwww: "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb bbbbbbbbbbbbbbbbbbbbbbb bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
                                             wwwwwwwwwwwwwwwwwwwwwwwwxw: "ccc bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb bbbbbbbbbbbbbbbbbbbbbbb bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb" },

                                                                          hide_keys: [],
                                                                              title: "Summary SUM",
                                                                              transform: :newline)

```

![bildschirmfoto 2017-01-11 um 20 54 17](https://cloud.githubusercontent.com/assets/2891702/21864033/2b616936-d840-11e6-90ef-03fd0193cb79.png)




and i also added support for this in the `actions_list.rb`:



```
fastlane actions
```

![bildschirmfoto 2017-01-11 um 20 55 15](https://cloud.githubusercontent.com/assets/2891702/21864053/4af99822-d840-11e6-841b-5f1e7d538a3b.png)



and in detailed action descr.:

```
fastlane action adb
```


![bildschirmfoto 2017-01-11 um 20 56 22](https://cloud.githubusercontent.com/assets/2891702/21864086/6f545bbc-d840-11e6-8727-1b68765907f3.png)

jinjorge commented:
is there a way to make the line wraps happen at word boundaries?

from the screenshot above there are several words that are broken up from one line to the next

hjanuschka commented:
@jinjorge  good point, but i think there are cases where really long words would not be wrapped and resulting in a ugly table, you have a idea on how to solve that?

jinjorge commented:
@hjanuschka It's a numbers game - there are a lot more sub 5-6 letter words than really long words. Probably best to optimize for the majority and see how things fall out.

To you question on how to solve it - I don't know, though I will do some research.

hjanuschka commented:
test's pass now, ready for new review! @jinjorge  i am unsure on how to solve it without breaking the output.

jinjorge commented:
a regular expression match (`\b`) for word boundary.

hjanuschka commented:
@jinjorge  got it working !!!  respecting whole word, but if too long forcing to break it! ;)

<img width="291" alt="bildschirmfoto 2017-01-24 um 08 03 29" src="https://cloud.githubusercontent.com/assets/2891702/22237410/e7e7a6fe-e20b-11e6-9e62-59d21deaa287.png">


```ruby
  # Base taken from: https://www.ruby-forum.com/topic/57805
  def wordwrap(length  = 80)
     self.gsub!(/(\S{#{length}})(?=\S)/, '\1 ')
     self.scan(/.{1,#{length}}(?:\s+|$)/)
  end
```

jinjorge commented:
👏 👏

milch commented on fastlane_core/lib/fastlane_core/print_table.rb:
```diff
@@ -34,16 +41,61 @@ def print_values(config: nil, title: nil, hide_keys: [], mask_keys: [])
+        return_array = []
+        tcols = IO.console.winsize[1]
+
+        col_count = rows.map(&:length).first || 1
```
Why is 1 being used as default here?

hjanuschka replied:
i used to use 0 - but well we have always atleast one column, otherwise it crashes later on due to a division by zero 💃

milch commented on fastlane_core/lib/fastlane_core/print_table.rb:
```diff
@@ -34,16 +41,61 @@ def print_values(config: nil, title: nil, hide_keys: [], mask_keys: [])
+            value = col.to_s.dup
+            if transform == :truncate_middle
+              value = value.middle_truncate(max_value_length)
+            end
```
Use `elsif` here to more clearly show intent

hjanuschka replied:
good point, going to adress this once back in private life 👍

hjanuschka replied:
ahhh sorry to fast reply, i think there are cases where transform is `false`  (e.g: CI run)
so its neither :newline nor :truncate_middle - any other ideas?

milch reviewed (dismissed):
Looks good to me aside from a couple minor things. I really think we should merge this soon

hjanuschka commented:
@milch  feedback addressed, as we have the super awesome full-word split now i enabled `:newline` by default, and therefore the `elsif` worked out pretty good 👍

DanToml commented on fastlane_core/lib/fastlane_core/print_table.rb:
```diff
@@ -24,7 +24,14 @@ def print_values(config: nil, title: nil, hide_keys: [], mask_keys: [])
 
         params = {}
-        params[:rows] = limit_row_size(rows)
+        transform = false if FastlaneCore::Env.truthy?("SKIP_TRANSFORM_TABLES")
```
`FL_SKIP_TABLE_TRANSFORM`?

hjanuschka replied:
done

DanToml commented on fastlane_core/lib/fastlane_core/print_table.rb:
```diff
@@ -34,16 +41,60 @@ def print_values(config: nil, title: nil, hide_keys: [], mask_keys: [])
 
-        max_key_length = rows.map { |e| e[0].length }.max || 0
-        max_allowed_value_length = max_length - max_key_length - 7
         rows.map do |e|
```
`rows.map do |row|`? - single letter variable made it pretty difficult to infer when scan reading

hjanuschka replied:
true point, fixed

DanToml commented on fastlane_core/lib/fastlane_core/print_table.rb:
```diff
@@ -34,16 +41,60 @@ def print_values(config: nil, title: nil, hide_keys: [], mask_keys: [])
+        terminal_table_padding = 4
+        max_length = tcols - (col_count * terminal_table_padding)
+
+        max_value_length = (max_length / col_count)
```
you don't need parens here

hjanuschka replied:
can we keep them?, i like it a bit more from point of readability.

DanToml replied:
I tend to only prefer them when there are multiple expressions to make precedence easier to read, but sure.

DanToml commented on fastlane_core/lib/fastlane_core/print_table.rb:
```diff
@@ -34,16 +41,60 @@ def print_values(config: nil, title: nil, hide_keys: [], mask_keys: [])
+          new_row = []
+          e.each do |col|
+            value = col.to_s.dup
+            if transform == :truncate_middle
```
would be nice to break this out into a transform_row/column function, because then appending to the row will be in a single place and be easier to maintain and test

i.e
```
def transform_row(transform, row)
  row.map do |column|
    transform_column(transform, column)
  end
end
```

```
def transform_column(transform, column)
  if transform == :truncate_middle
    ...
    value
  elseif transform == :newline
    ...
    value
end
```

hjanuschka replied:
i am not 100% comfortable with this, as there is bunch a logic going on in :newline, colorization for example - lets keep it as it is, and try to finish this PR - as its a really long running one, and do the modularization in a seperate PR. (i just broke way too much with last minute quick-changes :sad:, last PRs)

endocrimes reviewed (dismissed):
I like where this is going :+1:

hjanuschka reviewed (commented):
feedback adressed, master rebased and tty-screen used 👍

DanToml commented on fastlane_core/lib/fastlane_core/print_table.rb:
```diff
@@ -24,7 +24,14 @@ def print_values(config: nil, title: nil, hide_keys: [], mask_keys: [])
 
         params = {}
-        params[:rows] = limit_row_size(rows)
+        transform = false if FastlaneCore::Env.truthy?("FL_SKIP_TABLE_TRANSFORM")
```
rather than overloading this, I'd maybe extract out a `transform?` method that is `!FastlaneCore::Env.truthy?...` and have the if statement be `if transform? ...` - this means you don't need to trace through mutating state.

DanToml commented on fastlane_core/lib/fastlane_core/print_table.rb:
```diff
@@ -34,16 +41,60 @@ def print_values(config: nil, title: nil, hide_keys: [], mask_keys: [])
+
+        rows.map do |row|
+          new_row = []
+          row.each do |col|
```
as we only return one value from this, we could do `new_row = row.map do |col|`

DanToml replied:
might also be easier to read if we break the contents of this block into a `transform_column(column, truncation_style)` method?

KrauseFx replied:
updated 👍

giginet commented on fastlane_core/lib/fastlane_core/print_table.rb:
```diff
@@ -34,16 +41,60 @@ def print_values(config: nil, title: nil, hide_keys: [], mask_keys: [])
 
-      def limit_row_size(rows, max_length = 100)
+      def colorize_array(array, colors)
+        value = ""
```
you can write this logic more clearly.

```ruby
array.map { |l| "#{colors.first[0]}#{l}#{colors.last[0]}" unless colors.empty? }.join('\n')
```

mfurtak replied:
I think this may not be equivalent, because it can create `nil` entries

mfurtak commented:
Just checking in on the status of this PR - can the feedback be addressed? Everyone seems pretty positive about getting it in! 👍

KrauseFx commented:
Improving all fastlane tables is definitely something we should fix, it's a big issue currently. @hjanuschka, is that something you want to finish up? 🚀

hjanuschka commented:
@KrauseFx yeah i want to see this in.
however i worked alot on this, and the open feedback, is are just nit-picks, and i am afraid i break everything to resolve the feedback (which as a personal point of view, is not mission critical in this case)

so, how about merging this, and later on the feedback suggesters go on with it?

hjanuschka commented:
@googlebot thx for this, i accidentally commited with corporate email - as soon as we know how to proceed with this PR i am going to rebase then it should be fine 👍

hjanuschka commented:
master rebased, cla stuff sorted out.

KrauseFx commented on fastlane_core/lib/fastlane_core/print_table.rb:
```diff
@@ -34,16 +41,60 @@ def print_values(config: nil, title: nil, hide_keys: [], mask_keys: [])
       end
 
-      def limit_row_size(rows, max_length = 100)
+      def colorize_array(array, colors)
```
This method is never called, should it be part of this PR?

KrauseFx replied:
Whops, never mind, it is used

KrauseFx commented on fastlane_core/lib/fastlane_core/print_table.rb:
```diff
@@ -34,16 +41,60 @@ def print_values(config: nil, title: nil, hide_keys: [], mask_keys: [])
+            new_row << value
+          end
+          return_array << new_row
+          return_array << :separator
```
Why do we have this here? This will result in a new line of dashes inbetween every row, making the table a lot bigger

KrauseFx commented:
<img width="1039" alt="screen shot 2017-04-06 at 7 06 57 pm" src="https://cloud.githubusercontent.com/assets/869950/24782621/396e8a64-1afc-11e7-9a1a-6864e9681956.png">

I think not having the separators for every row looks better 👍

KrauseFx commented on fastlane_core/spec/print_table_spec.rb:
```diff
@@ -48,15 +48,15 @@
       value = FastlaneCore::PrintTable.print_values(config: @config, title: title, hide_keys: [:a_sensitive])
       expect(value[:title]).to eq(title.green)
-      expect(value[:rows]).to eq([['cert_name', "asdf"], ['output', '..'], ["a_bool", true]])
+      expect(value[:rows]).to eq([["cert_name", "asdf"], :separator, ["output", ".."], :separator, ["a_bool", "true"]])
```
It's clearly visible that the `separator` changes the existing behavior

KrauseFx commented:
Moved to https://github.com/fastlane/fastlane/pull/8809 (by mistake), I addressed all of your feedback, thanks. Also some more tests and code style changes.
